### PR TITLE
Catch error, parse it into expected model

### PR DESCRIPTION
**Pull request by Vilsol**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export function generateUUIDv4() {
+  return ("" + 1e7 + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, (c: any) =>
+    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
```
Comment (aaomidi) [inline src/utils.ts]: Nit. Does this need to be cryptographically secure? Just slows you down a bit.

File: src/models.ts [hunk 2]
```diff
@@ -19,6 +19,9 @@ export interface Environment {
 
   // Screen height
   screenHeight: number;
+
+  // The platform of the machine
+  platform: string;
```
Comment (aaomidi) [inline src/models.ts]: Platform feels ambiguous, can you add an example to the comment

File: src/index.ts [hunk 3]
```diff
@@ -1 +1,77 @@
...
+    screenHeight: window.screen.availHeight,
+    userAgent: window.navigator.userAgent,
+    windowWidth: window.innerWidth,
+    windowHeight: window.innerHeight,
+    url: window.location.href,
+    platform: window.navigator.platform
+  };
+}
+
+(<any> window)['InitializeErrorTracker'] = InitializeErrorTracker;
```
Comment (aaomidi) [inline src/index.ts]: Use window as any, not <any> window

File: src/index.ts [hunk 4]
```diff
@@ -1 +1,77 @@
...
+      } as ErrorStackTrace;
+    } else {
+      return undefined;
+    }
+  }).filter(valid => !!valid) as ErrorStackTrace[];
+}
+
+function collectEnvironmentData(): Environment {
+  return {
+    locale: window.navigator.language,
```
Comment (aaomidi) [inline src/index.ts]: What if any of these are 'undefined'. Say I put this in Node for example to catch server side errors?

File: src/index.ts [hunk 5]
```diff
@@ -1 +1,77 @@
...
+      return {
+        row: parseInt(parsed[3], 10),
+        col: parseInt(parsed[4], 10),
+        function: parsed[1],
+        file: parsed[2]
+      } as ErrorStackTrace;
+    } else {
+      return undefined;
+    }
+  }).filter(valid => !!valid) as ErrorStackTrace[];
```
Comment (aaomidi) [inline src/index.ts]: Just do boolean(valid) !! can have unintended side effects

File: src/index.ts [hunk 6]
```diff
@@ -1 +1,77 @@
...
+    string: stack
+  };
+}
+
+function parseErrorStackTrace(stack: string): ErrorStackTrace[] {
+  return stack.split('\n').slice(1).map((line) => {
+    const parsed = stackTraceLineRegex.exec(line);
+    if (parsed) {
+      return {
+        row: parseInt(parsed[3], 10),
```
Comment (aaomidi) [inline src/index.ts]: Check length of array before using it

File: src/index.ts [hunk 7]
```diff
@@ -1 +1,77 @@
...
+
+function parseErrorTrace(stack: string): ErrorTrace {
+  return {
+    stack: parseErrorStackTrace(stack),
+    string: stack
+  };
+}
+
+function parseErrorStackTrace(stack: string): ErrorStackTrace[] {
+  return stack.split('\n').slice(1).map((line) => {
```
Comment (aaomidi) [inline src/index.ts]: Double check '\n' You might even be able to just infer from the stack

File: src/index.ts [hunk 8]
```diff
@@ -1 +1,77 @@
...
+const reportVersion = 1;
+
+let sessionId: string;
+
+const stackTraceLineRegex = /^\s+at\s(\S+)\s\((.+):(\d+):(\d+)\)$/;
+
+export default function InitializeErrorTracker() {
+  sessionId = generateUUIDv4();
+
+  window.addEventListener("error", (error: ErrorEvent) => {
```
Comment (aaomidi) [inline src/index.ts]: FYI I recommend forcing an error to pop up after this listener as a "self-check" and report back if it works. Some other applications take over the entire error event system.

Also, allow people to attach their own error reporter to this. Provide a function they can call with an ErrorEvent.

File: src/utils.ts (hunk 1 above)
Comment (Vilsol) [reply src/utils.ts]: It gives it a bit more chance to be unique, and considering this is done once on page initialization it's not much of a burden.

File: src/index.ts (hunk 4 above)
Comment (Vilsol) [reply src/index.ts]: This is a browser-only tracker, there will be a separate project "nodejs-tracker" designed for server side.

File: src/index.ts (hunk 6 above)
Comment (Vilsol) [reply src/index.ts]: The array comes from a regex, it is bounded by start and end, so it cannot have fewer than 5 elements.

File: src/index.ts (hunk 8 above)
Comment (Vilsol) [reply src/index.ts]: Doing a self-check can be harmful if there are other error reporters, as they will not ignore the error.

Comment (aaomidi) [reply src/index.ts]: That makes sense, I would like to have a flag I can use at least once to check.

Comment (Vilsol) [reply src/index.ts]: You can just throw any error and see if it gets processed and logged or sent over the network

**Pull request merged**